proposals/p3848.md

```diff
@@ -0,0 +1,576 @@
+
+```
+fn Foo() {
+  let lambda: auto = fn[T:! Printable](t: T) { Print(t); };
+
+  fn FunctionDeclaration[T:! Printable](t: T) { Print(t); }
+}
+```
+
+To understand how the syntax between lambdas and function declarations is
+reasonably "continuous", refer to this table of syntactic positions and the
+following code examples.
+
+| Syntactic Position |                        Syntax Allowed in Given Position (optional, unless otherwise stated)                        |
+| :----------------: | :----------------------------------------------------------------------------------------------------------------: |
+|         A1         |               Required Returned Expression ([positional parameters](#positional-parameters) allowed)               |
+|         A2         |             Required Returned Expression ([positional parameters](#positional-parameters) disallowed)              |
+|         B          |                                   [Default Capture Mode](#default-capture-mode)                                    |
+|         C          | Explicit [Captures](#function-captures), [Function Fields](#function-fields) and Deduced Parameters (in any order) |
+|         D          |                                                Explicit Parameters                                                 |
+|         E1         |           Body of Statements (no return value) ([positional parameters](#positional-parameters) allowed)           |
+|         E2         |          Body of Statements (with return value) ([positional parameters](#positional-parameters) allowed)          |
+|         E3         |         Body of Statements (no return value) ([positional parameters](#positional-parameters) disallowed)          |
+|         E4         |        Body of Statements (with return value) ([positional parameters](#positional-parameters) disallowed)         |
+|         F          |                                                Required Return Type                                                |
+|         G          |                                             Function Declaration Name                                              |
+
+```
+// Lambdas (all the following are in an expression context and are
+// themselves expressions)
+
+fn => A1
+
+fn[B, C] => A1
+
+fn(D) => A2
+
+fn[B, C](D) => A2
+
+fn { E1; }
+
+fn -> F { E2; }
+
+fn[B, C] { E1; }
+
+fn[B, C] -> F { E2; }
+
+fn(D) { E3; }
+
+fn(D) -> F { E4; }
+
+fn[B, C](D) { E3; }
+
+fn[B, C](D) -> F { E4; }
+```
+
+```
+// Function Declarations (all the following are allowed as statements in a
+// function body or as declarations in other scopes)
+
+fn G => A1;
+
+fn G[B, C] => A1;
+
+fn G(D) => A2;
+
+fn G[B, C](D) => A2;
+
+fn G { E1; }
+
+fn G -> F { E2; }
+
+fn G[B, C] { E1; }
+
+fn G[B, C] -> F { E2; }
+
+fn G(D) { E3; }
+
+fn G(D) -> F { E4; }
+
+fn G[B, C](D) { E3; }
+
+fn G[B, C](D) -> F { E4; }
+```
+
+**Alternative Considered**: As opposed to a continuous syntax between lambdas
```

Comment:
We often link to relevant alternatives from the relevant design docs. In order to make them linkable, I think they need to be marked with ### section headings.

Reply:
Had to employ embedded HTML tags to make this work. Hopefully that's ok.

Reply:
Unfortunately when I ran precommit on this, it removed these html tags from the table of contents, even though I manually added them. The sections are still linkable, they just don't show up in the TOC unfortunately.